lib/chatops/controller.rb

```diff
@@ -52,6 +49,46 @@ def execute_chatop
 
     protected
 
+    def setup_params
+      json_body.each do |key, value|
+        next if params.has_key? key
+        params[key] = value
+      end
+
+      permitted_params = %i[
+        action
+        chatop
+        controller
+        mention_slug
+        method
+        room_id
+        user
+      ]
+
+      chatop_name =
+        if params[:chatop].present?
+          params[:chatop].to_sym
+        elsif params[:action].present?
+          params[:action].to_sym
+        else
+          nil
+        end
+
+      if chatop = self.class.chatops[chatop_name]
+        permitted_params << { params: chatop[:params] }
+      end
+
+      self.params = params.permit(*permitted_params)
+    end
+
+    def json_body
+      hash = {}
+      if request.content_type =~ %r/\Aapplication\/json\Z/i
+        hash = GitHub::JSON.parse(request.raw_post) || {}
+      end
+      hash.with_indifferent_access
+    end
+
     def jsonrpc_params
```

Comment:
It looks like you've left this method in, but it's now only referenced by the tests. Given that this exposes access to data within the `"params"` key, not at the root of the params, does that mean you need to permit params as being `params: permitted_params`?

Reply:
> does that mean you need to permit params as being `params: permitted_params`?

That is correct. The code that takes care of that permitting is here: 
https://github.com/github/chatops-controller/blob/450b7c5f947d569cf7bda3d7622e4110eb8b8459/lib/chatops/controller.rb#L76-L78

It is using the named captures from the chatop regular expression to identify what JSON RPC params the current chatop method can accept. Those named captures become the list of permitted JSON RPC params.

The `jsonrpc_params` method is the main entry point used by chatop implementations to access the JSON RPC chatop arguments passed in by hubot when a chatop is executed. So that method should definitely stick around.